{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code to save svgs with a defined size and font"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# For pheatmap\n",
    "pheatmap(heatmap.combo, filename = paste(fig_dir, 'cell_class_heatmap.pdf', sep='/'), width = 3.4, height = 1.6, fontsize = 7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# For ggsave\n",
    "ggsave(gautier_UMAP, filename = paste(fig_dir, 'gautier_UMAP.svg', sep='/'), device='svg', width = 3, height = 3, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# For pheatmap, width = 3.2 in, height 4 in or less\n",
    "# For scatter plots, width = height = 1 in\n",
    "# Font = 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“The following package(s) are missing entries in the cache:\n",
      "\n",
      "\tremotes\n",
      "\n",
      "These packages will need to be reinstalled.\n",
      "”\n"
     ]
    }
   ],
   "source": [
    "renv::init('/oak/stanford/groups/agitler/Shared/Shared_Jupyter_Notebook_Analysis/4.1.1-OG/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:hdf5r’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gplots’:\n",
      "\n",
      "    space\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "library(ggrepel)\n",
    "#library(ComplexHeatmap)\n",
    "library(pheatmap)\n",
    "library(gplots)\n",
    "library(svglite)\n",
    "library(hdf5r)\n",
    "library(DropletUtils)\n",
    "library(scDblFinder)\n",
    "library(scuttle)\n",
    "library(scales)\n",
    "library(ggrastr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘future’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:hdf5r’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(future)\n",
    "plan(\"multicore\", workers = availableCores())\n",
    "options(future.globals.maxSize= +Inf)\n",
    "# NotebookApp.max_buffer_size=214748364800"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File Loading and Saving Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "load.dir <- '/oak/stanford/groups/agitler/Shared_Sequencing/Jacob/Levine-Human-FastQ/h5_files'\n",
    "fig_dir <- '/oak/stanford/groups/agitler/Shared/Matters_Arising/Figure_Panels'\n",
    "csv_dir <- '/oak/stanford/groups/agitler/Shared/Matters_Arising/CSV_Files'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load H5 Matrices from Yadav 2022"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x = counts[]), :\n",
      "“'giveCsparse' has been deprecated; setting 'repr = \"T\"' for you”\n",
      "Warning message in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x = counts[]), :\n",
      "“'giveCsparse' has been deprecated; setting 'repr = \"T\"' for you”\n",
      "Warning message in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x = counts[]), :\n",
      "“'giveCsparse' has been deprecated; setting 'repr = \"T\"' for you”\n",
      "Warning message in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x = counts[]), :\n",
      "“'giveCsparse' has been deprecated; setting 'repr = \"T\"' for you”\n",
      "Warning message in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x = counts[]), :\n",
      "“'giveCsparse' has been deprecated; setting 'repr = \"T\"' for you”\n",
      "Warning message in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x = counts[]), :\n",
      "“'giveCsparse' has been deprecated; setting 'repr = \"T\"' for you”\n",
      "Warning message in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x = counts[]), :\n",
      "“'giveCsparse' has been deprecated; setting 'repr = \"T\"' for you”\n"
     ]
    }
   ],
   "source": [
    "donor1_h5 <- Read10X_h5(paste(load.dir,'GSM5723843_Donor1_raw_feature_bc_matrix.h5',sep='/'), use.names = TRUE, unique.features = TRUE)\n",
    "donor2_h5 <- Read10X_h5(paste(load.dir,'GSM5723844_Donor2_raw_feature_bc_matrix.h5',sep='/'), use.names = TRUE, unique.features = TRUE)\n",
    "donor3_h5 <- Read10X_h5(paste(load.dir,'GSM5723845_Donor3_raw_feature_bc_matrix.h5',sep='/'), use.names = TRUE, unique.features = TRUE)\n",
    "donor4_h5 <- Read10X_h5(paste(load.dir,'GSM5723846_Donor4_raw_feature_bc_matrix.h5',sep='/'), use.names = TRUE, unique.features = TRUE)\n",
    "donor5_h5 <- Read10X_h5(paste(load.dir,'GSM5723847_Donor5_raw_feature_bc_matrix.h5',sep='/'), use.names = TRUE, unique.features = TRUE)\n",
    "donor6_h5 <- Read10X_h5(paste(load.dir,'GSM5723848_Donor6_raw_feature_bc_matrix.h5',sep='/'), use.names = TRUE, unique.features = TRUE)\n",
    "donor7_h5 <- Read10X_h5(paste(load.dir,'GSM5723849_Donor7_raw_feature_bc_matrix.h5',sep='/'), use.names = TRUE, unique.features = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bcrank1 <- barcodeRanks(donor1_h5)\n",
    "bcrank2 <- barcodeRanks(donor2_h5)\n",
    "bcrank3 <- barcodeRanks(donor3_h5)\n",
    "bcrank4 <- barcodeRanks(donor4_h5)\n",
    "bcrank5 <- barcodeRanks(donor5_h5)\n",
    "bcrank6 <- barcodeRanks(donor6_h5)\n",
    "bcrank7 <- barcodeRanks(donor7_h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "debris_donor1 <- donor1_h5[,donor1_h5['NEFH', ] >1 & bcrank1$total<2000 & bcrank1$total>100]\n",
    "donor1.debris.seurat <- CreateSeuratObject(debris_donor1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "debris_donor2 <- donor2_h5[,donor2_h5['NEFH', ] >1 & bcrank2$total<2000 & bcrank2$total>100]\n",
    "donor2.debris.seurat <- CreateSeuratObject(debris_donor2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "debris_donor3 <- donor3_h5[,donor3_h5['NEFH', ] >1 & bcrank3$total<2000 & bcrank3$total>100]\n",
    "donor3.debris.seurat <- CreateSeuratObject(debris_donor3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "debris_donor4 <- donor4_h5[,donor4_h5['NEFH', ] >1 & bcrank4$total<2000 & bcrank4$total>100]\n",
    "donor4.debris.seurat <- CreateSeuratObject(debris_donor4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "debris_donor5 <- donor5_h5[,donor5_h5['NEFH', ] >1 & bcrank5$total<2000 & bcrank5$total>100]\n",
    "donor5.debris.seurat <- CreateSeuratObject(debris_donor5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "debris_donor6 <- donor6_h5[,donor6_h5['NEFH', ] >1 & bcrank6$total<2000 & bcrank6$total>100]\n",
    "donor6.debris.seurat <- CreateSeuratObject(debris_donor6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "debris_donor7 <- donor7_h5[,donor7_h5['NEFH', ] >1 & bcrank7$total<2000 & bcrank7$total>100]\n",
    "donor7.debris.seurat <- CreateSeuratObject(debris_donor7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Log Normalize list of Seurat Objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "source('/oak/stanford/groups/agitler/Jacob/single_cell_motor_neurons/seurat_analysis_scripts/sc_functions.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"splitting complete\"\n",
      "[1] 1\n",
      "[1] \"normalization complete\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -3.0007”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.30103”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  3.025e-15”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"variable feature selection complete\"\n",
      "[1] 2\n",
      "[1] \"normalization complete\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -2.7132”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.32138”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  7.9239e-29”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“There are other near singularities as well. 0.090619”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"variable feature selection complete\"\n",
      "[1] 3\n",
      "[1] \"normalization complete\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -2.3959”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.32036”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  6.7313e-28”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“There are other near singularities as well. 0.090619”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"variable feature selection complete\"\n",
      "[1] 4\n",
      "[1] \"normalization complete\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -2.6375”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.32149”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  3.9454e-29”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“There are other near singularities as well. 0.090619”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"variable feature selection complete\"\n",
      "[1] 5\n",
      "[1] \"normalization complete\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -2.0897”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.31887”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  2.6703e-28”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“There are other near singularities as well. 0.031008”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"variable feature selection complete\"\n",
      "[1] 6\n",
      "[1] \"normalization complete\"\n",
      "[1] \"variable feature selection complete\"\n",
      "[1] 7\n",
      "[1] \"normalization complete\"\n",
      "[1] \"variable feature selection complete\"\n"
     ]
    }
   ],
   "source": [
    "debris.objects.list <- c(donor1.debris.seurat, donor2.debris.seurat, donor3.debris.seurat, donor4.debris.seurat, donor5.debris.seurat, donor6.debris.seurat, donor7.debris.seurat)\n",
    "debris.objects.list.normalized <- normalize.aggregated.object(debris.objects.list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove Mitochondrial Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for(i in 1:length(debris.objects.list.normalized)) {\n",
    "    counts <- GetAssayData(debris.objects.list.normalized[[i]], assay = \"RNA\")\n",
    "    mito_genes <- grep( \"^MT-\", rownames(debris.objects.list.normalized[[i]]), value = T)\n",
    "    counts <- counts[-(which(rownames(counts) %in% mito_genes)),]\n",
    "    debris.objects.list.normalized[[i]] <- subset(debris.objects.list.normalized[[i]], features = rownames(counts))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find Variable Features etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -3.0007”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.30103”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  1.2356e-14”\n",
      "Loading required package: foreach\n",
      "\n",
      "Loading required package: iterators\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ARHGAP24, KIAA1328, PLXDC2, LINC01608, LINC00862, RNF219-AS1, CLMN, C3, GRID2, CDK19 \n",
      "\t   PCDH9-AS2, MIR2052HG, RORB, ZNF131, GRIK1, APBB1IP, HIF1A-AS2, NEK11, ZHX2, FAM66D \n",
      "\t   BCL2L14, AOAH, ARHGAP15, MEF2C, PDE7A, RGCC, ELMO1-AS1, LINC01184, FOXO1, ZNF862 \n",
      "Negative:  NEFM, NEFL, MAP1B, HSP90AA1, TUBB2A, UCHL1, HSP90AB1, STMN2, NEFH, PRPH \n",
      "\t   KIF21A, YWHAG, PKM, ATP1B1, VAT1L, RTN4, THY1, RTN3, ACTG1, AKAP12 \n",
      "\t   TUBB4B, DYNC1H1, RTN1, SV2C, CLU, TUBA1B, SNCG, GAPDH, NDRG4, ENO2 \n",
      "PC_ 2 \n",
      "Positive:  NEFH, PRPH, ANXA2, UTS2, S100A10, FBXO2, SLC5A7, MCAM, CPNE6, TUBGCP2 \n",
      "\t   ACTG1, HSPB1, LGALS1, TPPP3, VAT1L, RAB3D, MGST3, CHODL, NEFL, PAPPA \n",
      "\t   DYNLL1, ACLY, PVALB, YWHAZ, CALCA, CLCN5, ATP1A1, SOD1, PRUNE2, CEND1 \n",
      "Negative:  VSNL1, INA, LAMP5, SLC6A5, SLC17A6, ATP2B2, HAPLN4, ANK1, IQSEC1, VSTM2L \n",
      "\t   ATP2B3, SLC12A5, GABBR2, RAB22A, TMEM130, FAM19A2, FAM85B, SCN1A, ZFHX3, PWAR6 \n",
      "\t   ZMAT4, AC244021.1, STXBP1, SLC2A13, GRID2, MYO1F, STEAP2, AC008014.1, CACNA1D, LGI2 \n",
      "PC_ 3 \n",
      "Positive:  BICDL1, NINL, XKR6, CTNNBL1, SYNDIG1, APBB1IP, ZFHX3, CDR2, LINC01572, AC008014.1 \n",
      "\t   DOCK8, ERICH1, CPNE8, SYTL3, CSGALNACT1, MYO1F, FAM19A2, PPP1R16B, XIAP, AC104461.1 \n",
      "\t   MSANTD2, CCDC26, RTF1, AC244021.1, IQSEC3, NCOR2, ENOX2, KHDC4, ZNF337-AS1, CDH26 \n",
      "Negative:  INA, LAMP5, SLC17A6, SLC6A5, VSNL1, ATP2B2, DNM1, NEFH, HAPLN4, STXBP1 \n",
      "\t   LGI2, IQSEC1, SERINC1, SYT2, NPTN, RPH3A, SLC12A5, NSF, GABRB2, SCN1A \n",
      "\t   PLD3, PRNP, SCG3, PEG10, ATP2B3, PCYOX1, NCDN, SV2B, GRIN1, SNAP25 \n",
      "PC_ 4 \n",
      "Positive:  AC244021.1, MYO1F, FAM149A, APBB1IP, INPP5D, DOCK8, LHFPL2, SMAD3, ESR1, LYN \n",
      "\t   ABR, AP002495.2, P2RY12, CCDC26, ARHGAP22, DLEU7-AS1, CACNA1D, NAAA, ZFHX3, MIR646HG \n",
      "\t   CATSPER2, C3, ST6GAL1, MGAT4A, ARHGAP15, RNASEH2B, LRRK1, MEF2C, KCNQ3, KHDC4 \n",
      "Negative:  SYN3, ABCD2, FAM85B, NCOR2, CIT, RMC1, VSTM2L, PRICKLE1, IQSEC3, CEP250 \n",
      "\t   ZNF337-AS1, TMEM132B, NINL, SPDYE3, AC104461.1, AL139294.1, CTNNBL1, CDR2, CUX2, AC100800.1 \n",
      "\t   PPP1R16B, TAF4, NR4A3, ANO6, CPNE8, SRRM4, LINC02381, CDH26, AC008014.1, ENOX2 \n",
      "PC_ 5 \n",
      "Positive:  AC105402.3, RAB6B, MTRNR2L12, DHFR, EBF1, PWAR6, RGS7BP, SLC45A4, RAB22A, PAK3 \n",
      "\t   C14orf132, IDS, KCNB1, KCNC2, DDX52, PRKAA2, MAP1A, NUAK1, AC025682.1, NID1 \n",
      "\t   RAB3C, PTGER3, COMMD8, ROCK2, SSU72, MTRNR2L8, AC124312.1, FAM98A, SORBS2, ELAVL2 \n",
      "Negative:  SLC6A5, TMEM132B, PRICKLE1, FAM66D, RMC1, ZNF891, ATP2B3, ANO6, NEFH, CIT \n",
      "\t   CDR2, CEP250, ATP6AP2, RFWD3, LRRK2, ZNF337-AS1, MPHOSPH9, SPARCL1, SYNDIG1, CDH26 \n",
      "\t   CTNNBL1, ZMAT4, AC092683.1, LPCAT2, ST6GAL1, PVALB, LINC01572, PPHLN1, THOC7, DDX55 \n",
      "\n",
      "Warning message:\n",
      "“The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session”\n",
      "16:33:57 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:33:57 Read 2003 rows and found 30 numeric columns\n",
      "\n",
      "16:33:57 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:33:57 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:33:58 Writing NN index file to temp file /tmp/RtmpMBkzmG/file64632048d9a9\n",
      "\n",
      "16:33:58 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:33:58 Annoy recall = 100%\n",
      "\n",
      "16:34:00 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:34:04 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:34:04 Commencing optimization for 500 epochs, with 95806 positive edges\n",
      "\n",
      "16:34:11 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 2003\n",
      "Number of edges: 142348\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.7667\n",
      "Number of communities: 3\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -2.7132”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.32138”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  4.3742e-28”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“There are other near singularities as well. 0.090619”\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  NEFM, NEFL, MAP1B, TUBB2A, HSP90AA1, STMN2, HSP90AB1, ATP1B1, PKM, RTN4 \n",
      "\t   ATP1A1, UCHL1, RTN1, PRPH, MAP7D2, YWHAG, SNAP25, SV2C, TUBB4B, KIF21A \n",
      "\t   CALM3, SLC5A7, AKAP12, SPTAN1, GOT1, UTS2, SEZ6L2, LGALS1, PRDX1, LDHB \n",
      "Negative:  IL13RA1, CNTNAP4, CDK13, KCNQ3, SLC9A9, PACS2, DLG2, XIST, PPP4R3A, SPARC \n",
      "\t   STPG2, GNAI3, NCAM2, CNOT9, SMURF1, PPP2R2B, SPTLC3, DOCK4, FANCC, SLMAP \n",
      "\t   AC012409.2, DLEU2, TMCC1, FAM198B-AS1, MDGA2, CPLANE1, HDAC9, ANGPT1, PCSK5, AC034195.1 \n",
      "PC_ 2 \n",
      "Positive:  AC068888.1, IL4R, DNAJC15, AC004076.1, AC012409.2, COL14A1, ADAMTS9, NR2F2-AS1, IQGAP2, KDM3B \n",
      "\t   AC079447.1, XKR6, KCTD1, THBS1, AL109930.1, GCC2-AS1, CALD1, EXT1, PUM3, SH3PXD2A \n",
      "\t   COL4A2, CRISPLD2, RASGEF1C, ATP8B1, SLC36A1, PDLIM3, PRPF3, ACTN1, ATXN7L1, COL4A1 \n",
      "Negative:  NEFL, SLC25A32, HSP90AA1, EPHA4, HDAC3, CCDC184, SOX5, BAP1, RP2, KPNA4 \n",
      "\t   KIF5C, AL138828.1, LINC02362, EFCAB5, FHL1, QTRT2, NECTIN3, RALB, SRM, RERG \n",
      "\t   UBAC1, POLI, NEK5, TVP23B, RELL1, GABRA2, ZNF611, OST4, SST, B3GALT4 \n",
      "PC_ 3 \n",
      "Positive:  GRIA1, UNC5D, STK32B, UFL1-AS1, LIN28B, STX18-AS1, CPEB2-DT, ABLIM3, LINC01933, PGGT1B \n",
      "\t   ZMAT4, SLC36A1, FGD1, PEBP4, MARCH11, AL356124.1, SMIM18, MRAP2, RTN4IP1, AC090921.1 \n",
      "\t   PCBD2, AC131934.1, CASP3, TLR3, MRPL39, GABRG3, TRIP4, AC058791.1, SRRM3, CLVS2 \n",
      "Negative:  IQGAP2, NR2F2-AS1, AC012409.2, ADAMTS9, AC079447.1, FILIP1L, AL109930.1, IL4R, ATP8B1, RASGEF1C \n",
      "\t   PDLIM3, WWC3, BBOF1, PRRG4, TMEM131L, CACNA1C, AF241726.2, PRPF3, LINC00598, RNFT2 \n",
      "\t   DNAJC15, SH3PXD2A, RDH10-AS1, AC107068.2, CR381653.1, LTBP4, WFDC2, CDH23, LAMA4, AC005828.3 \n",
      "PC_ 4 \n",
      "Positive:  CTBP2, UNC13C, NABP2, TRMT61B, C21orf91, POMZP3, TYW1, ATG2B, INO80, HIPK2 \n",
      "\t   ESF1, AC239809.3, PTPRE, GRK5, DHX32, AC004448.2, FAM122C, AL035427.1, PIGB, DNAAF4 \n",
      "\t   DPP6, PRKRIP1, ARNTL, PYGO1, TOE1, CYYR1-AS1, TCERG1L, HMGN1, AL157832.3, GTF3C2 \n",
      "Negative:  RNFT2, LINC00598, GAS2L3, CR381653.1, LTBP4, CRLF1, GATC, ST3GAL6-AS1, WFDC2, GCC2-AS1 \n",
      "\t   AC005828.3, RDH10-AS1, AC107068.2, AC138123.1, PRRG4, BBOF1, AC005821.1, CDH23, THBS1, LGI4 \n",
      "\t   HIRA, COL8A1, CNOT6, ZSCAN16-AS1, H6PD, PRPF3, CRACR2A, TMC6, AC068888.1, RIOK2 \n",
      "PC_ 5 \n",
      "Positive:  COL4A2, ACTN1, CRISPLD2, COL4A1, JAG1, LINC00476, ADAMTS9, FRMD3, COL14A1, AC104316.1 \n",
      "\t   URM1, RADIL, RPGRIP1L, SHQ1, HDX, AP000919.2, LEKR1, AC004943.2, XYLB, RBPMS \n",
      "\t   RNF152, TIMP3, ZKSCAN8, AMZ1, COL15A1, AC012409.2, NHSL2, CDC37L1, JADE1, TAGLN \n",
      "Negative:  NABP2, ZBTB7A, C21orf91, GCC2-AS1, SLC35F4, AC092683.1, SH3PXD2A, KLHL29, UNC13C, L2HGDH \n",
      "\t   CDK13, DPP6, CDH4, SEC23IP, AL109930.1, RGMA, SORCS1, DHX32, PCLO, C21orf62-AS1 \n",
      "\t   AMPH, ANKRD36, EIPR1, ZNF589, UBE2E3, IQGAP2, HGF, LANCL2, GEN1, BACH1 \n",
      "\n",
      "16:34:15 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:34:15 Read 493 rows and found 30 numeric columns\n",
      "\n",
      "16:34:15 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:34:15 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:34:15 Writing NN index file to temp file /tmp/RtmpMBkzmG/file6463185568d5\n",
      "\n",
      "16:34:15 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:34:15 Annoy recall = 100%\n",
      "\n",
      "16:34:17 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:34:20 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:34:20 Commencing optimization for 500 epochs, with 18726 positive edges\n",
      "\n",
      "16:34:24 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 493\n",
      "Number of edges: 35524\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.6242\n",
      "Number of communities: 2\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -2.3959”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.32036”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  6.7313e-28”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“There are other near singularities as well. 0.090619”\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RPL34, EPB41L2, DENND5A, NEAT1, PREX2, GRIK2, RASGRP3, ATAD2B, HS6ST3, COP1 \n",
      "\t   LONP2, ZCCHC7, CSGALNACT2, TP53INP2, AL133346.1, FOXN3, YBX1, RAB5A, ZNF706, AC004158.1 \n",
      "\t   PTPN9, SVEP1, SH2B1, HDAC8, CD86, TIMM23B, GRAMD2B, CEBPG, CWF19L1, PFKFB2 \n",
      "Negative:  NEFM, NEFL, STMN2, EEF1A2, THY1, HSP90AB1, MAP1B, ATP1A1, UTS2, FAIM2 \n",
      "\t   PKM, HSP90AA1, REEP1, YWHAG, FKBP4, ATP1B1, PRNP, SRPRA, SYT2, FBXO2 \n",
      "\t   CLPTM1, SNCG, TUBA4A, ANXA6, UCHL1, SNAP25, KIF21A, CALM1, CHODL, MDH1 \n",
      "PC_ 2 \n",
      "Positive:  NEFH, NEFM, HSP90AA1, THY1, ACTG1, UTS2, KIF21A, VPS35, MDH1, RTN1 \n",
      "\t   TUBA1B, NEFL, ASB6, FKBP4, CLU, SNCG, SERINC1, UCHL1, ANXA6, SPOCK2 \n",
      "\t   ATP1B1, FBXO2, SESN3, SCG2, TUBA4A, FAIM2, CALY, SRPRA, SLC12A5, EGFL7 \n",
      "Negative:  RHBDF2, ADCY7, TMC8, MLXIPL, DNAJC5, DDX31, PRKAG2, IQGAP2, NLRC5, AC092920.1 \n",
      "\t   KCNQ1, EHBP1L1, LINC01001, DHRSX, MAN1A1, TRMT2A, INTS6L, INPP5D, LAIR1, CD163 \n",
      "\t   CRYBG1, CHST11, DENND3, SH3RF3, FAM157C, TBXAS1, CTCF, SEC24D, SLC7A5, KCNQ3 \n",
      "PC_ 3 \n",
      "Positive:  AC003991.1, HECW1, GNG12-AS1, RASSF5, PWP1, FAM153A, SDR16C5, LINC00535, WNK2, OSBPL10 \n",
      "\t   AC138123.1, GUCY1A1, PSMA5, GRIK3, SZRD1, AL445623.2, SLC8A3, C4orf19, C8orf34, AP000787.1 \n",
      "\t   GPR176, CA8, SEMA6D, NKAIN3, CLVS1, XKR4, LSM12, ARHGAP44, LRMP, DGKG \n",
      "Negative:  RHBDF2, MLXIPL, ADCY7, NLRC5, DNAJC5, TMC8, CLPTM1, DDX31, FAM157C, DENND3 \n",
      "\t   NEFH, INPP5D, SKI, NFIC, CRLF3, AC092920.1, LINC01001, TRNT1, EHBP1L1, PFKFB3 \n",
      "\t   INTS6L, NEFM, GATAD2A, TRMT2A, SH3RF3, TBC1D14, IQGAP2, UTRN, STAB1, ARRB2 \n",
      "PC_ 4 \n",
      "Positive:  IQGAP2, THEMIS, TMC8, CCDC88C, IKZF3, L3MBTL3, SKAP1, AOAH, PELI2, ATAD5 \n",
      "\t   AC009093.2, RBM4, ANKRD23, TC2N, RLF, AC080162.1, CRYBG1, AC027097.2, SLFN12L, DHRSX \n",
      "\t   STK4, EMB, MTMR6, GNPTAB, DERL2, STYX, PLA2G7, CTSC, LTA4H, CCND3 \n",
      "Negative:  GRAPL, EHD4, MLXIPL, CENPO, FLNB, PLCD1, PEX10, SH2D3C, HS6ST1, NRDE2 \n",
      "\t   KIF20B, CD151, TMEM132D, DAZAP1, SYT17, STK32B, MCM9, AC008591.1, CFAP44, WDFY4 \n",
      "\t   EHBP1L1, AC093157.1, NSUN6, MTA1, IFT57, FILIP1, ARRDC2, FCGR2A, OPALIN, DEK \n",
      "PC_ 5 \n",
      "Positive:  ADCY7, PREX1, TRMT2A, P2RX4, AC244131.2, PLA2G7, CEP295, HSPA6, KIAA0100, ARRB2 \n",
      "\t   KLHDC10, TBC1D9, NFIC, B4GALT1, ELP1, LURAP1L-AS1, AGAP6, MYO1F, RASSF3, AC015971.1 \n",
      "\t   HPN, RCOR1, HOTAIRM1, EIF5B, RUBCNL, AC011933.4, RREB1, GFPT1, XYLB, C8orf59 \n",
      "Negative:  AC092920.1, THEMIS, IKZF3, TC2N, CD69, ANKRD23, SKAP1, CCDC88C, ATXN2L, SLFN12L \n",
      "\t   AC009093.2, STYX, PELI2, AC116366.3, AC027097.2, SYNE2, MTFR1, AC105339.3, SYTL3, NLRC5 \n",
      "\t   INPP4A, STX3, CLEC16A, ID4, TOX, OFD1, TDG, OXNAD1, RIPOR2, GNPTAB \n",
      "\n",
      "16:34:26 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:34:26 Read 238 rows and found 30 numeric columns\n",
      "\n",
      "16:34:26 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:34:26 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:34:26 Writing NN index file to temp file /tmp/RtmpMBkzmG/file646317722f22\n",
      "\n",
      "16:34:26 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:34:26 Annoy recall = 100%\n",
      "\n",
      "16:34:28 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:34:32 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:34:32 Commencing optimization for 500 epochs, with 8306 positive edges\n",
      "\n",
      "16:34:34 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 238\n",
      "Number of edges: 17187\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.5301\n",
      "Number of communities: 2\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -2.6375”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.32149”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  3.9454e-29”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“There are other near singularities as well. 0.090619”\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  NEFL, NEFM, TUBB2A, PRPH, ATP1A1, PKM, MAP1B, CLU, HSP90AA1, STMN2 \n",
      "\t   RTN1, TUBA1B, HSP90AB1, ANXA2, ATP1B1, ACTG1, GAPDH, LGALS1, UCHL1, TUBB4B \n",
      "\t   RTN4, DYNC1H1, APLP2, SYT2, TUBA1A, PSAP, S100A10, CHODL, CADM3, UTS2 \n",
      "Negative:  LCORL, VRK2, CBWD5, PARG, FANCC, DOPEY1, FKBP14, GMDS, CRB1, DOCK2 \n",
      "\t   ALDH1L2, TANC2, LSAMP, BRINP3, SIDT1, AC073941.1, C2orf69, BCO2, SHOC2, SPAG1 \n",
      "\t   RBM8A, HIST2H2BF, MEF2C-AS1, HDAC9, ATF7IP2, AC002429.2, PHF6, KCNJ2, CACHD1, METTL15 \n",
      "PC_ 2 \n",
      "Positive:  ECM2, POLQ, ARHGAP26-AS1, DAGLB, FO393400.1, NMNAT3, GTF2IRD1, PIGU, OSBPL10, AL035427.1 \n",
      "\t   IL12RB2, AL033504.1, FBN2, SPDYE5, RBMS3-AS2, RSAD1, SAMD5, KIZ-AS1, ZNF667, CHCHD6 \n",
      "\t   ZBTB4, ZNF385D-AS2, EPHB1, GALNT17, NINL, NUP210, NEK10, EAF2, AC005162.3, ZNF777 \n",
      "Negative:  HSP90AA1, UCHL1, PRPH, TUBB2A, PKM, KIF21A, NEFL, CLU, UTS2, NEFM \n",
      "\t   RTN4, STMN2, HSP90AB1, CHODL, TUBA1B, TUBA1A, ACTG1, FBXO2, APLP2, ATP1A1 \n",
      "\t   AKAP12, CHGA, SHH, MCAM, SPTAN1, LEPROTL1, MAP7D2, USP7, CANX, TUBB4B \n",
      "PC_ 3 \n",
      "Positive:  HSPG2, NNMT, CLDN5, NR2F2, S1PR1, AHNAK, WDR78, CYP1B1, SORCS1, PDLIM1 \n",
      "\t   ALKBH3, MYH9, SCAF4, ZNF441, PECAM1, EPAS1, HYAL2, APOLD1, PTPN14, FTO \n",
      "\t   FOXJ3, METTL21A, PAPSS2, SLC7A5, ADAMTS1, SYNE2, MMRN1, AC109492.1, ACKR1, MIPOL1 \n",
      "Negative:  IL12RB2, KIZ-AS1, SPDYE5, ZNF385D-AS2, OSBPL10, SAMD5, NEK10, AL035427.1, AL033504.1, ZNF667 \n",
      "\t   FBN2, RBMS3-AS2, RSAD1, EPHB1, SLC22A4, GTF2IRD1, NUP210, KLHL3, ZNF385D, NPAT \n",
      "\t   ADCYAP1R1, EPHA5-AS1, AC005162.3, PIGU, FILIP1L, AL354771.1, CHCHD6, FGF9, DIRC2, FO393400.1 \n",
      "PC_ 4 \n",
      "Positive:  APOLD1, NR2F2, KLF2, AHNAK, NNMT, EPAS1, NTHL1, ACKR1, ID1, CLDN5 \n",
      "\t   ADAMTS1, VWF, HYAL2, WDR78, S1PR1, METTL21A, HLA-E, PAPSS2, ZNF441, LIPG \n",
      "\t   H2AFJ, ABHD5, MMRN1, CYP1B1, FXYD5, AC109492.1, PDLIM1, HSPG2, MYH9, SYNE2 \n",
      "Negative:  LST1, NKAPL, CHEK2, MCM5, C1orf112, AC026202.2, CLSPN, ARRB1, AC004943.2, KCNMB1 \n",
      "\t   TMC8, THAP6, KBTBD12, EED, EME2, ACAD9, ARRB2, CTNND1, BLNK, AQP11 \n",
      "\t   IL1RAP, AC090825.1, P2RY12, FOXP2, CCDC24, LIN9, SYTL3, METTL16, EZH2, FAM129A \n",
      "PC_ 5 \n",
      "Positive:  SLC7A5, PGBD5, ALKBH3, SPHKAP, KSR1, AL109930.1, EMC8, AL358075.2, BANP, NAT10 \n",
      "\t   MEGF11, TSHZ2, TESMIN, BZW2, ATG4A, DLGAP2, ZBTB4, FEM1B, NNMT, UBR4 \n",
      "\t   ZCCHC14, POLR1C, SIPA1L2, ITGA2B, CENPP, ZNF586, RCC1L, HCN1, NMNAT3, PFDN1 \n",
      "Negative:  GSE1, CSPG4, PAPPA, SHC4, ADCY9, GAS2L3, AC010997.3, NR2F2-AS1, COL27A1, LGI4 \n",
      "\t   HSPG2, CMTM7, NES, FBXO15, NDST3, CAMK1, LRRC40, UQCC3, RCAN1, LINC00598 \n",
      "\t   SH3RF1, TRMT13, ZNF438, ARHGEF28, MORN4, NTNG1, RELB, SORCS1, SCN7A, LINC00511 \n",
      "\n",
      "16:34:37 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:34:37 Read 414 rows and found 30 numeric columns\n",
      "\n",
      "16:34:37 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:34:37 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:34:37 Writing NN index file to temp file /tmp/RtmpMBkzmG/file646367d84110\n",
      "\n",
      "16:34:37 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:34:37 Annoy recall = 100%\n",
      "\n",
      "16:34:39 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:34:43 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:34:43 Commencing optimization for 500 epochs, with 14922 positive edges\n",
      "\n",
      "16:34:46 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 414\n",
      "Number of edges: 33445\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.5879\n",
      "Number of communities: 2\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“pseudoinverse used at -2.0897”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“neighborhood radius 0.31887”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“reciprocal condition number  9.4779e-28”\n",
      "Warning message in simpleLoess(y, x, w, span, degree = degree, parametric = parametric, :\n",
      "“There are other near singularities as well. 0.031008”\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ZNF98, CACNG2, IQCA1, TNK2, PDZD7, ASCC2, THOC5, ELL, SULT1A1, BCR \n",
      "\t   GOLGA7B, ADGRL1, THRA, TOM1, LRRC57, TMEM38A, TSHZ3, ZNF208, RUSC1, DLL1 \n",
      "\t   CRKL, UNC13A, TMEM120B, RAB11FIP4, DNAJC7, KIF1A, RUNDC3A, PNKD, CACNA1A, EIF3L \n",
      "Negative:  MEF2A, LRRC4C, TANC2, NBEA, UBAC2, SMURF1, ULK4, FAM49B, CNOT2, NPAS2 \n",
      "\t   XIST, SRP54, PKP4, CLASP2, MGA, AL392023.2, LAMP2, MDGA2, CRADD, RBM41 \n",
      "\t   CNTNAP5, SMYD3, LCLAT1, ZNF251, TM2D3, AC124312.1, ZBED5, RNF125, MTMR2, KCTD9 \n",
      "PC_ 2 \n",
      "Positive:  IQCA1, CACNG2, ZNF98, GATAD2A, TSHZ3, TMEM38A, LRRFIP1, PDZD7, JAZF1, DACH1 \n",
      "\t   ZNF208, THOC5, ASCC2, CRKL, TNK2, BRD4, RAB11FIP4, THRA, EIF3L, ZNF714 \n",
      "\t   UNC13A, GOLGA7B, CUL3, KLHL12, CUX1, TEFM, FAM66C, SCAI, SFI1, AP1G1 \n",
      "Negative:  THY1, STMN2, NEFM, URB1, NEFL, EEF1A2, YWHAG, RETREG3, MCAM, UTS2 \n",
      "\t   GNB4, ATP6V0E2, TMEM120A, SPARCL1, NFE2L1, TMX4, PRMT7, OTUB1, SPOCK2, SLC2A1 \n",
      "\t   ATP1B1, CLU, MAP1B, EXOSC9, TUBB2A, CCT6A, PODXL2, CALY, CALM1, HSP90AB1 \n",
      "PC_ 3 \n",
      "Positive:  ARF1, DYNLL1, ARNTL, KRBOX4, B4GAT1, API5, AC123768.3, GOLGA4, EMSY, PGRMC2 \n",
      "\t   STOM, E4F1, ZNF423, PCBP2, ALOX5, TRANK1, XRCC6, BCL2L13, ATP6V1E1, AP3D1 \n",
      "\t   MBNL3, CPM, HINT1, CTBP2, FP671120.1, H2AFY2, KCTD10, PTPRJ, EIF5B, UBE2G1 \n",
      "Negative:  ATP6V0E2, TOM1, ZDBF2, SLC25A4, EEF1A2, ABCA10, C17orf67, PRMT7, STMP1, PODXL2 \n",
      "\t   FAM227B, ZNF565, NPIPB5, ANXA6, CACNA1D, RIMS2, GPALPP1, LRAT, SLC6A5, USP53 \n",
      "\t   IFI16, CYB5R4, WASHC3, CBX1, SCFD2, ACYP1, PIBF1, AC006059.1, ZFPM2, KLF13 \n",
      "PC_ 4 \n",
      "Positive:  ZSWIM3, SPTLC1, PITHD1, CCT6A, SLC2A1, EMC1, RAB3A, KIF3A, BCL6, CCT3 \n",
      "\t   DNMBP, VMP1, GTF2F2, SEMA3B, HKR1, GYG1, DTWD1, GOLGA3, TXNL1, AP1B1 \n",
      "\t   LINC02210-CRHR1, UHRF1BP1L, CTR9, SLC2A3, HECTD4, UGP2, FDX1, LRRN1, MEGF11, HNRNPH3 \n",
      "Negative:  CLHC1, NELL2, TSPAN14, NIPSNAP1, ZNF487, KRBOX4, EEF1A2, NCDN, PDIA3, PRMT7 \n",
      "\t   H2AFY2, SF3B3, ARF1, EIF3J, SLC16A10, PLXNA4, DISC1, ATP6V0E2, TRIM26, GTF3C3 \n",
      "\t   DMXL2, CLIP3, SNORC, SPECC1, RHBDL3, GPATCH1, LYST, STAT5B, ITGB4, TOM1 \n",
      "PC_ 5 \n",
      "Positive:  SAFB2, HDAC6, FP671120.1, E4F1, ZNF423, U2SURP, PRELID2, FAM174B, TMEM70, GABRA3 \n",
      "\t   GPATCH2L, CTBP2, LMBR1, SIN3B, EMSY, GAR1, CYLD, KCNB1, BCAN, WAPL \n",
      "\t   RPL36, ZNF131, PINK1, MANEA, CNBP, FBXO42, BNIP3, CHST12, AL139021.1, SDR39U1 \n",
      "Negative:  DNAJC10, AC123768.3, PIGL, LINC00461, ZNF827, NCDN, CDR2, RRAGB, TKT, SLC4A10 \n",
      "\t   GOLIM4, ZNF487, KLHL3, ZNF680, SKA2, CLHC1, KPNA5, KLHL42, RGS3, CA3 \n",
      "\t   UTP6, CCT6A, ATXN3, TRIM26, ZNF439, MAP3K1, FOXN2, TSEN15, RAB5B, HEATR5B \n",
      "\n",
      "16:34:48 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:34:48 Read 118 rows and found 30 numeric columns\n",
      "\n",
      "16:34:48 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:34:48 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:34:48 Writing NN index file to temp file /tmp/RtmpMBkzmG/file646344b6dbc3\n",
      "\n",
      "16:34:48 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:34:48 Annoy recall = 100%\n",
      "\n",
      "16:34:50 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:34:53 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:34:53 Commencing optimization for 500 epochs, with 3780 positive edges\n",
      "\n",
      "16:34:56 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 118\n",
      "Number of edges: 6899\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.5000\n",
      "Number of communities: 1\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  NEFM, NEFL, NEFH, HSP90AA1, MAP1B, UCHL1, HSP90AB1, PKM, THY1, TUBB2A \n",
      "\t   STMN2, CLU, PRPH, KLC1, ATP1A1, ATP1B1, ENO2, EEF1A2, YWHAG, RTN1 \n",
      "\t   RTN3, SV2C, TUBA1B, SNAP25, ANXA6, NDRG4, VAT1L, CADM3, GAPDH, TUBB4B \n",
      "Negative:  NRG3, MIR99AHG, PLEK, C10orf90, DOCK4, SGCZ, FKBP5, SLC1A3, ETV6, ARHGAP26 \n",
      "\t   KCNQ3, FAM193A, KIAA0586, KIRREL3, ENOX1, LRMDA, VAV1, CEMIP, MIR646HG, MTRNR2L8 \n",
      "\t   RNF220, INPP5D, LDLRAD4, RUNX1, PEAK1, DHRS3, IL17RA, DISC1, PREX1, SLC2A5 \n",
      "PC_ 2 \n",
      "Positive:  NEFH, NRXN1, H2AFZ, NEFM, ZNF793, RNF220, SCN8A, TRDN, MAP1B, GTF3A \n",
      "\t   SARAF, ZBTB16, RTN1, FGF12, NRG3, RTN4, LANCL1-AS1, HS6ST3, AC103876.1, ATP2B2 \n",
      "\t   SELENOS, AC066613.2, SV2A, MB21D2, ATP1A1, RFX4, FGF14, PPP2R1A, FSIP2, DNAAF4 \n",
      "Negative:  PRAM1, TNFRSF1B, SH3RF3, PALD1, ABR, IGSF21, CD83, PIK3R5, MYO1F, ITGAX \n",
      "\t   MLXIPL, NFATC2, CHST11, CYFIP1, ELL, INPP5D, KCNQ3, CYTH4, SLCO2B1, IKZF1 \n",
      "\t   FP671120.1, VAV1, C22orf34, C3, PFKFB3, CSF2RA, B4GALT1, NR4A3, FRMD4A, KCNIP1 \n",
      "PC_ 3 \n",
      "Positive:  PRPH, ANXA2, NEFH, CHODL, MCAM, CPNE6, S100A10, LGALS1, SLC5A7, TMSB10 \n",
      "\t   SOD1, FBXO2, TSC22D3, DYNLL1, BASP1, HTRA1, HSPB1, SUSD2, UTS2, GAP43 \n",
      "\t   HINT1, SPP1, MAP1A, HSPB6, FXYD7, VAT1L, CCDC184, SNCG, CD151, YWHAH \n",
      "Negative:  LAMP5, INA, VSNL1, SLC6A5, SLC17A6, GRIN1, SLC32A1, ATP2B2, HAPLN4, DISP2 \n",
      "\t   SYT2, DNM1, MEGF8, SLC22A17, SV2A, CKB, CLSTN1, NSF, STXBP1, KCNA2 \n",
      "\t   ATP6AP2, NDUFAF5, CHGB, TGFBI, ASAH2, LGI3, NXPH4, SYP, SEZ6L2, HOXA7 \n",
      "PC_ 4 \n",
      "Positive:  PRAM1, INA, NEFH, IKZF1, SLC6A5, CYTH4, SH3RF3, MYO1F, HOXA7, RCOR1 \n",
      "\t   MLXIPL, FAM149A, RGS12, C3, FCGBP, CLU, PALD1, MDH1, TNFRSF1B, ATP1B1 \n",
      "\t   SLC17A6, ITGAX, NEFM, AHCY, HAPLN4, MEF2D, SLC25A5, PRELID1, EIF5A2, RASSF8 \n",
      "Negative:  AJAP1, POLN, LETM1, B3GALT5, GRIK3, SYN3, CACNA1E, RHBDL3, ANKRD30BL, ATAD3B \n",
      "\t   EBF4, FAM193A, MAMDC2-AS1, EBF3, CACNG2, UBR4, CHD5, TIMP3, WNK2, ARHGEF10L \n",
      "\t   COL19A1, EBF1, DENND5B, GRIK1, CEP104, MYO16, PDXK, GRIA1, OAS1, LARP1 \n",
      "PC_ 5 \n",
      "Positive:  NPPC, IPO9-AS1, GRIA1, SYT4, NETO2, SYN3, CACNG2, COL19A1, SPTB, DOK5 \n",
      "\t   LAMA3, CACNA1E, MAMDC2-AS1, LINC01933, B3GALT5, ATP8B2, MIEF1, DARS2, MRPS18A, EBF1 \n",
      "\t   MYO16, PDYN, GRIK1, CACNA1G, EBF4, LIMK2, DGKI, RGS17, OGDHL, EBF3 \n",
      "Negative:  DIRC3, ARHGEF10L, DSN1, PTDSS2, CSMD2, MIR4435-2HG, MORC4, ATAD3B, CFAP46, LINC02381 \n",
      "\t   PAX2, DSTYK, PAQR9, EBPL, ZNF451-AS1, NPHP4, SLC9A1, RUNDC3B, MRNIP, RHBDL3 \n",
      "\t   PTPRN, CEP104, TRPC5, KIF1A, KIF9, EPHA4, SLITRK4, RPL37A, FLNA, SLC16A10 \n",
      "\n",
      "16:35:00 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:35:00 Read 1841 rows and found 30 numeric columns\n",
      "\n",
      "16:35:00 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:35:00 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:35:01 Writing NN index file to temp file /tmp/RtmpMBkzmG/file646358756b94\n",
      "\n",
      "16:35:01 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:35:01 Annoy recall = 100%\n",
      "\n",
      "16:35:03 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:35:07 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:35:07 Commencing optimization for 500 epochs, with 84394 positive edges\n",
      "\n",
      "16:35:14 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 1841\n",
      "Number of edges: 118140\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.7368\n",
      "Number of communities: 4\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  MALAT1, NEFH, ST18, SLC44A1, FRMD4B, LRMDA, IL1RAPL1, GFAP, GAB1, DOCK10 \n",
      "\t   FCER1G, S100A11, ZEB2, ARL15, TCF12, FTX, ZBTB20, SAMSN1, ITPR2, VRK2 \n",
      "\t   KCNH8, AC092691.1, PDE1A, TBXAS1, EFEMP1, CD69, FMNL2, UNC5C, CNDP1, SRGN \n",
      "Negative:  ATP1B1, INA, HSP90AA1, MAP1B, NSF, VSNL1, STXBP1, NEFM, DNM1, ANK2 \n",
      "\t   SYT2, CLSTN1, GABRG2, SLC6A5, SCN1A, LAMP5, ATP2B2, SLC22A17, NEFL, SV2A \n",
      "\t   PEG10, CKB, HAPLN4, OXR1, CAMTA1, SLC32A1, IDS, SCG2, CALB2, USP11 \n",
      "PC_ 2 \n",
      "Positive:  INA, MALAT1, VSNL1, SLC6A5, LAMP5, SV2A, SLC32A1, SLC17A6, GABBR2, NSG2 \n",
      "\t   SLC8A1, PGRMC1, ATP2B2, GAD1, GAD2, KCNC1, GPRASP1, GABRB2, CALB2, GABRA1 \n",
      "\t   OPRM1, KCNA2, BTBD8, PCP4, SYT1, STEAP2, LY6H, KCNC2, HAPLN4, CELF4 \n",
      "Negative:  PRPH, NEFL, NEFM, SPP1, UTS2, MAP1B, CHODL, SLC5A7, NEFH, SPARCL1 \n",
      "\t   HSP90AA1, PSD3, NRGN, SPTBN1, LIFR, MAP1A, NRG1, CALCA, MAP2, PGM2L1 \n",
      "\t   HCN1, PLCB1, NELL2, ARFGEF1, GOLGB1, SLC35F3, HSD17B12, DHX36, NCL, PAK3 \n",
      "PC_ 3 \n",
      "Positive:  LINC00513, MEG8, MIAT, GRIP1, MEG3, MALAT1, AC092683.1, SGCD, LINC00632, OSBP2 \n",
      "\t   SGCZ, SLC44A5, AC124312.1, ZHX3, PWAR6, GRID1, TTC28, CDH12, MIR29B2CHG, DAB1 \n",
      "\t   DCC, DLGAP2, SLC35F4, AC079760.1, MTMR3, AC016831.7, KCNQ1OT1, DROSHA, CAPS2, DENND1B \n",
      "Negative:  NEFH, HSP90AA1, NEFM, NEFL, ATP1B1, MAP1B, NSF, DNM1, LAMP5, SLC6A5 \n",
      "\t   INA, SYT2, STXBP1, SLC22A17, SCN1A, CLSTN1, GABRG2, CALB2, HAPLN4, USP11 \n",
      "\t   SPP1, CKB, VSNL1, OXR1, SPARCL1, SV2A, PEG10, ATP2B2, ANK2, SLC32A1 \n",
      "PC_ 4 \n",
      "Positive:  PWAR6, DHFR, AC105402.3, AC124312.1, MTRNR2L12, MAP2, IDS, MAP1A, CCNI, NUDT3 \n",
      "\t   PAK3, AL078639.1, MAPK10, ZNF106, SACS, RBMS3, ADCY1, TTBK2, ZNF483, ADRA1A \n",
      "\t   CSRNP3, RALB, MAPK8IP2, SPTBN1, BICD1, MEG8, GNG4, PTGER3, PSD3, GUCY1A2 \n",
      "Negative:  NEFH, HSP90AA1, LINC00513, NSF, KCNH8, KIRREL3, LINC01608, SPP1, DNM1, DCC \n",
      "\t   ANK2, TMTC2, ATP1B1, DNM3, SCMH1, MYO1D, ZNF536, LINC00882, CDKL1, MALAT1 \n",
      "\t   TCF12, MIR181A1HG, OXR1, DSCAML1, ST18, SPOCK3, PDE1A, UVRAG, SCN1A, NEFM \n",
      "PC_ 5 \n",
      "Positive:  ATG10, MYO9B, DOCK2, ELL2, ZEB2, MAP1A, CCDC7, ZNF331, LYN, BNC2 \n",
      "\t   NCK2, PTK2B, MCC, FAM49A, PTPN1, NPL, DISC1, SIPA1L2, MIER2, MKLN1 \n",
      "\t   CCDC26, MSR1, INPP5D, PAG1, ETV6, ZSWIM6, REL, NR4A2, LARP7, DOCK4 \n",
      "Negative:  MIAT, CABIN1, AC092683.1, LINGO2, MEG3, MTMR3, TTC28, GRIP1, MEG8, LIMK2 \n",
      "\t   AC239809.3, ZNF280B, SLIT2, SEZ6L, MKL1, FAM153A, FAM227A, LINC02231, CDH12, SGSM1 \n",
      "\t   SGCZ, SLC35F4, NEFH, SLC44A5, DLGAP2, KIAA1671, PCDH15, FAM85B, CACNG2, AC127526.1 \n",
      "\n",
      "16:35:21 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:35:21 Read 3433 rows and found 30 numeric columns\n",
      "\n",
      "16:35:21 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:35:21 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:35:22 Writing NN index file to temp file /tmp/RtmpMBkzmG/file6463376b9455\n",
      "\n",
      "16:35:22 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:35:22 Annoy recall = 100%\n",
      "\n",
      "16:35:24 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:35:28 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:35:28 Commencing optimization for 500 epochs, with 153120 positive edges\n",
      "\n",
      "16:35:41 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 3433\n",
      "Number of edges: 174123\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.7920\n",
      "Number of communities: 6\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 singletons identified. 5 final clusters.\n",
      "\n",
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n"
     ]
    }
   ],
   "source": [
    "for(i in 1:length(debris.objects.list.normalized)) {\n",
    "    DefaultAssay(debris.objects.list.normalized[[i]]) <- 'RNA'\n",
    "    debris.objects.list.normalized[[i]] <- FindVariableFeatures(debris.objects.list.normalized[[i]])\n",
    "    debris.objects.list.normalized[[i]] <- find.pcs.and.clustids(debris.objects.list.normalized[[i]])\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Integrate Across Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CheckDuplicateCellNames(object.list = object.list):\n",
      "“Some cell names are duplicated across objects provided. Renaming to enforce unique cell names.”\n",
      "Computing 2000 integration features\n",
      "\n",
      "Scaling features for provided objects\n",
      "\n",
      "Finding all pairwise anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2441 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1576 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1189 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1115 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1155 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1089 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2061 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1425 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1744 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1200 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1147 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 808 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 590 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "Warning message in FilterAnchors(object = object.pair, assay = assay, slot = slot, :\n",
      "“Number of anchor cells is less than k.filter. Retaining all anchors.”\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 590 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "Warning message in FilterAnchors(object = object.pair, assay = assay, slot = slot, :\n",
      "“Number of anchor cells is less than k.filter. Retaining all anchors.”\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 578 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "Warning message in FilterAnchors(object = object.pair, assay = assay, slot = slot, :\n",
      "“Number of anchor cells is less than k.filter. Retaining all anchors.”\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 590 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "Warning message in FilterAnchors(object = object.pair, assay = assay, slot = slot, :\n",
      "“Number of anchor cells is less than k.filter. Retaining all anchors.”\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 7036 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1776 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2436 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 650 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1190 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 302 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2056 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 564 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 590 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "Warning message in FilterAnchors(object = object.pair, assay = assay, slot = slot, :\n",
      "“Number of anchor cells is less than k.filter. Retaining all anchors.”\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 8803 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1725 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2463 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 479 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1190 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 212 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2069 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 358 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 590 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "Warning message in FilterAnchors(object = object.pair, assay = assay, slot = slot, :\n",
      "“Number of anchor cells is less than k.filter. Retaining all anchors.”\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 8298 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1356 anchors\n",
      "\n"
     ]
    }
   ],
   "source": [
    "debris.droplets.anchors <- FindIntegrationAnchors(object.list = debris.objects.list.normalized, )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Merging dataset 5 into 1\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 3 into 1 5\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 4 into 1 5 3\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 2 into 1 5 3 4\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 6 into 1 5 3 4 2\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 7 into 1 5 3 4 2 6\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "debris.droplets.integrated <- IntegrateData(anchorset = debris.droplets.anchors, verbose=T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  NEAT1, DLG2, LRP1B, ERBB4, NRG3, NCAM2, DLC1, DPP10, DOCK4, MIR99AHG \n",
      "\t   RNF220, MAML2, GPC5, SGCZ, CNTNAP2, LRMDA, KLHL32, ANKS1B, LRRC4C, FAM155A \n",
      "\t   KIRREL3, RBFOX1, FRMD4A, SAMD12, ENOX1, NRXN1, ARHGAP26, GALNT13, LAMA2, TTC28 \n",
      "Negative:  NEFL, NEFM, HSP90AB1, HSP90AA1, UCHL1, NEFH, TUBB2A, STMN2, MAP1B, HSPA8 \n",
      "\t   PKM, ATP1B1, CALM1, S100A10, DYNC1H1, GNAS, PRPH, MAP7D2, DNAJA1, VAT1L \n",
      "\t   AKAP12, ANXA2, HSPB1, AHNAK2, PVALB, UTS2, MAP9, NSF, PLD3, DNM1 \n",
      "PC_ 2 \n",
      "Positive:  MAP1B, ATP1B1, PKM, NEFM, NSF, HSP90AA1, DYNC1H1, SYT2, HSP90AB1, DNM3 \n",
      "\t   PLD3, CLSTN1, MAP7D2, DNM1, NEFL, DPP6, SCN1A, TUBB2A, STXBP1, UCHL1 \n",
      "\t   GABRG2, AKAP12, RBFOX1, STMN2, NRXN1, VAT1L, NCDN, GNAS, CALM1, CADPS \n",
      "Negative:  NEFH, AC105402.3, GFAP, HBB, S100A11, DHFR, MTRNR2L12, HIF1A-AS2, AC024230.1, HBA2 \n",
      "\t   VAMP8, AL358216.1, AC024257.1, AC015923.1, DHDH, AL353768.1, AC022613.1, SULT1C4, LPAR4, AC138956.2 \n",
      "\t   GPC2, MPZ, AC007114.1, SLC13A5, AC003666.1, AL359317.2, LINC00463, AP000345.2, LGALS2, AC068533.4 \n",
      "PC_ 3 \n",
      "Positive:  PRPH, S100A10, ANXA2, UTS2, SLC5A7, CHODL, VAT1L, HSPB1, NEFH, NEFL \n",
      "\t   AHNAK2, NEFM, TUBB2A, MAP1B, AKAP12, STMN2, NRGN, PVALB, SLIT3, DNAJA1 \n",
      "\t   CALCA, UCHL1, DYNC1H1, SPARCL1, LIFR, PAPPA, MAP1A, HCN1, SEMA3A, PSMA7 \n",
      "Negative:  INA, VSNL1, LAMP5, SLC6A5, SV2A, SLC32A1, SLC17A6, ATP2B2, CALB2, GABBR2 \n",
      "\t   KCNC1, NSG2, HAPLN4, SLC22A17, KCNA2, GAD2, DNM1, GPRASP1, STXBP1, GABRA1 \n",
      "\t   PCP4, GAD1, GABRB2, CKB, CLSTN1, NCDN, STEAP2, CELF4, CDS1, OPRM1 \n",
      "PC_ 4 \n",
      "Positive:  SH3RF3, INPP5D, MYO1F, TNFRSF1B, IKZF1, PALD1, CYFIP1, ABR, NFATC2, CHST11 \n",
      "\t   SLCO2B1, TNS3, KCNQ3, DENND3, C3, PIK3R5, FLI1, CD83, DISC1, SLC11A1 \n",
      "\t   PRAM1, DOCK2, RUNX1, AOAH, ST6GAL1, AC244021.1, RHBDF2, KCNIP1, IPCEF1, HS3ST4 \n",
      "Negative:  FGF14, PWAR6, AC124312.1, SLC35F3, KAZN, DPP10, MIAT, NRG3, DLG2, DIP2C \n",
      "\t   CNTNAP2, DNM3, GRIP1, NCAM2, SLC35F4, PAK3, CDH18, MAP2, NRG1, RFX4 \n",
      "\t   CDH12, LRP1B, GRM7, KIRREL3, DLC1, GRID1, ROBO2, KCNIP4, FAM155A, RALYL \n",
      "PC_ 5 \n",
      "Positive:  PWAR6, AC105402.3, DHFR, AC124312.1, MTRNR2L12, MAP2, IDS, MAP1A, PAK3, CLVS2 \n",
      "\t   AL078639.1, DOK6, ADRA1A, CALM1, CACNG2, NRG1, GUCY1A2, DCTN5, MYT1L, DENND1B \n",
      "\t   RFPL1S, ADCY1, GRIN2B, RALB, GNG4, MIR4435-2HG, AC244021.1, MIR34AHG, SLIT3, NRSN1 \n",
      "Negative:  NEFH, HSP90AA1, PKM, NSF, RNF220, ERBB4, DNM1, DLC1, MAP7D2, NEFM \n",
      "\t   NCAM2, AC026316.5, DNM3, LRP1B, PLD3, SLC6A5, NRXN1, LAMP5, DLG2, RNF219-AS1 \n",
      "\t   ATP1B1, SPOCK3, CLSTN1, GPC5, MIR99AHG, USP11, SPARCL1, HSP90AB1, NRG3, CADM1 \n",
      "\n",
      "16:38:00 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:38:00 Read 8540 rows and found 30 numeric columns\n",
      "\n",
      "16:38:00 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:38:00 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:38:01 Writing NN index file to temp file /tmp/RtmpMBkzmG/file64636dd37a80\n",
      "\n",
      "16:38:01 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:38:02 Annoy recall = 100%\n",
      "\n",
      "16:38:04 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:38:07 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:38:08 Commencing optimization for 500 epochs, with 404544 positive edges\n",
      "\n",
      "16:38:34 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 8540\n",
      "Number of edges: 357615\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.8605\n",
      "Number of communities: 9\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n"
     ]
    }
   ],
   "source": [
    "DefaultAssay(debris.droplets.integrated) <- 'RNA'\n",
    "debris.droplets.integrated <- FindVariableFeatures(debris.droplets.integrated)\n",
    "debris.droplets.integrated <- find.pcs.and.clustids(debris.droplets.integrated)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Based on expression of NEAT1 and NEFM, Clusters 2, 4, and 5 are neuritic debris"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DefaultAssay(debris.droplets.integrated) <- 'integrated'\n",
    "debris.droplets.integrated.final <- subset(debris.droplets.integrated, idents=c(2,3,5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in FindVariableFeatures.Assay(object = assay.data, selection.method = selection.method, :\n",
      "“selection.method set to 'vst' but count slot is empty; will use data slot instead”\n",
      "Warning message in eval(predvars, data, env):\n",
      "“NaNs produced”\n",
      "Warning message in hvf.info$variance.expected[not.const] <- 10^fit$fitted:\n",
      "“number of items to replace is not a multiple of replacement length”\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  MALAT1, LINC02082, ZNF665, AC079465.1, AC080162.1, E2F5, AC008074.1, AL031280.1, PRRX1, AL133480.1 \n",
      "\t   ANO2, LINC01482, AC009690.1, C10orf90, AC007495.2, KIRREL3, TRG-AS1, IL12A-AS1, AC104596.1, NFKB1 \n",
      "\t   RAD54L2, SOX5, ECM2, RHBDF2, NAMPT, KAT6A, AC008758.4, SYCE2, BMPR1B, DIAPH3 \n",
      "Negative:  NEFM, NEFL, MAP1B, HSP90AA1, NEFH, HSP90AB1, UCHL1, RTN4, GAPDH, RTN1 \n",
      "\t   STMN2, YWHAG, RTN3, CLU, TUBB2A, SNAP25, DYNC1H1, SV2C, PRNP, AKAP12 \n",
      "\t   KIF21A, ACTG1, ENO2, MAP7D2, EPB41L3, PKM, TUBB4B, PRPH, VAT1L, ACLY \n",
      "PC_ 2 \n",
      "Positive:  PRPH, TRG-AS1, S100A10, ZNF665, SPP1, NEFL, NEFH, SLC5A7, ANXA2, UTS2 \n",
      "\t   MCAM, SLFN12L, AC021231.1, DCLK1, AC079465.1, ACTG1, DOCK8, PRDM6, AP002495.2, GSE1 \n",
      "\t   AC018742.1, CHODL, AC109466.1, PDGFD, MFSD1, PXN, IGFBP7, LINC01933, TUBA1A, CHST11 \n",
      "Negative:  INA, SLC6A5, VSNL1, LAMP5, SLC17A6, AC104041.1, SLC32A1, ATP2B2, DENND3, SCN1A \n",
      "\t   SV2A, KCNC1, NSF, HAPLN4, ITGB4, SLC22A17, STXBP1, CPXM2, AC107068.2, CLSTN1 \n",
      "\t   GAD1, SCN8A, LGI2, ABAT, ATP1B1, MIPEP, AC019130.1, ANK1, PSAP, AL357153.2 \n",
      "PC_ 3 \n",
      "Positive:  AC080162.1, EIF3L, RUNX1, AC008074.1, PRRX1, DCX, RBFOX1, ACOT9, AL445250.1, AL356124.1 \n",
      "\t   SPRY4-AS1, PRDM6, ST6GAL1, AC007495.2, NAMPT, VIPR2, AL049840.1, PRR11, ELMOD1, IDS \n",
      "\t   OSMR-AS1, CLPB, HOOK3, PROX1, PAK3, ERC2, AC008758.4, ARHGAP15, MAU2, C3orf58 \n",
      "Negative:  LINC01933, E2F5, AC079465.1, AP002495.2, CEMIP2, AC018742.1, LINC00310, AC009975.1, AC026316.5, AC091078.1 \n",
      "\t   RASGEF1B, AC005154.6, HMGN5, SH3RF1, CASC2, LINC01122, CFAP47, COL8A1, ZNF277, AC021231.1 \n",
      "\t   VPS13B, NIPSNAP1, TMEM117, MARK1, HOMER1, ASXL1, DTWD2, NUP214, SCMH1, CATSPER2 \n",
      "PC_ 4 \n",
      "Positive:  AL356124.1, NEFH, NEFM, AL031280.1, NEFL, E2F5, HSP90AA1, UCHL1, CLU, NARS \n",
      "\t   NSF, HSPA12A, MANBA, SLC6A5, LINC02240, TUBB2A, SPP1, AL139383.1, TANC2, KLC1 \n",
      "\t   PCED1B, PKM, STK35, MYO1D, ECM2, LDHB, TC2N, CAPRIN2, SERINC1, SNCG \n",
      "Negative:  EBF1, MTRNR2L12, AC091078.1, NEXMIF, CPLANE1, PURG, KCNJ6, RYBP, TRG-AS1, POLQ \n",
      "\t   AC124312.1, COL19A1, AC108488.2, HIVEP3, FGF14, PWAR6, XIST, LINC01933, ZBTB7A, ROBO1 \n",
      "\t   PGGT1B, FAM155A, EXT1, C18orf25, PGAP1, KCNC2, AC008537.1, SLC12A6, TMEM178B, AC244021.1 \n",
      "PC_ 5 \n",
      "Positive:  AC108488.2, MTRNR2L12, AC124312.1, NEXMIF, KCNC2, AL031280.1, ARL5B, LINC02082, PWAR6, TSHZ2 \n",
      "\t   PAPSS2, PLCXD3, ZNF490, ZBTB7A, NEK5, FAM19A2, SH3RF1, FRAS1, MFHAS1, MAP1A \n",
      "\t   SACS, GALNT13, EBF1, DHFR, RBFOX2, LYST, PAK3, AC105402.3, VGLL3, NRG1 \n",
      "Negative:  GRIP1, LINC01933, PALM2, ZNF487, LINC00310, SCMH1, AL035401.1, AC080162.1, AC008074.1, APP \n",
      "\t   BACH1, SGCD, RNF220, LINC02362, OSMR-AS1, DCX, STAG1, MALAT1, DOCK2, TTC17 \n",
      "\t   COL19A1, ROBO1, ZNF714, AC019131.1, AC119868.2, LINC01250, NAV3, LRP1B, DENND3, EXT1 \n",
      "\n",
      "16:38:42 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:38:42 Read 2908 rows and found 30 numeric columns\n",
      "\n",
      "16:38:42 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:38:42 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:38:43 Writing NN index file to temp file /tmp/RtmpMBkzmG/file64632c758a16\n",
      "\n",
      "16:38:43 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:38:43 Annoy recall = 100%\n",
      "\n",
      "16:38:45 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:38:48 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:38:48 Commencing optimization for 500 epochs, with 124140 positive edges\n",
      "\n",
      "16:38:58 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 2908\n",
      "Number of edges: 185073\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.8236\n",
      "Number of communities: 7\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n"
     ]
    }
   ],
   "source": [
    "DefaultAssay(debris.droplets.integrated) <- 'RNA'\n",
    "debris.droplets.integrated.final <- FindVariableFeatures(debris.droplets.integrated.final)\n",
    "debris.droplets.integrated.final <- find.pcs.and.clustids(debris.droplets.integrated.final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DefaultAssay(debris.droplets.integrated.final) <- 'integrated'\n",
    "debris.droplets.integrated.neurites <- subset(debris.droplets.integrated.final, idents=c(0,1,2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in FindVariableFeatures.Assay(object = assay.data, selection.method = selection.method, :\n",
      "“selection.method set to 'vst' but count slot is empty; will use data slot instead”\n",
      "Warning message in eval(predvars, data, env):\n",
      "“NaNs produced”\n",
      "Warning message in hvf.info$variance.expected[not.const] <- 10^fit$fitted:\n",
      "“number of items to replace is not a multiple of replacement length”\n",
      "PC_ 1 \n",
      "Positive:  NEFL, NEFM, NEFH, MAP1B, PRPH, UCHL1, ACTG1, HSP90AA1, VAT1L, MCAM \n",
      "\t   ANXA2, GAPDH, TUBB2A, HSP90AB1, STMN2, CLU, MAP7D2, RTN1, S100A10, RTN4 \n",
      "\t   LGALS1, ENO2, KLC1, SLC5A7, KIF21A, ACLY, KIF3A, AKAP12, EPB41L3, DYNC1H1 \n",
      "Negative:  MALAT1, E2F5, LINC02082, AC008074.1, AC079465.1, LINC01482, AL133480.1, DIAPH3, IL12A-AS1, ZNF331 \n",
      "\t   GALNTL6, RGS22, C10orf90, RHBDF2, KAT6A, AC009690.1, TOX3, AL031280.1, VSNL1, NLGN1 \n",
      "\t   AC104596.1, ZNF665, RNF220, AL139294.1, TC2N, ANO2, SOX5, COBLL1, AC009975.1, SUFU \n",
      "PC_ 2 \n",
      "Positive:  INA, SLC6A5, LAMP5, VSNL1, AC104041.1, SLC32A1, NSF, ATP1B1, SCN1A, ATP2B2 \n",
      "\t   SLC22A17, CLSTN1, ITGB4, SERINC1, HAPLN4, STXBP1, DENND3, SLC17A6, SCN8A, AC107068.2 \n",
      "\t   KCNC1, SV2A, CPXM2, CLTC, PSAP, LGI2, AC019130.1, MIPEP, GAD1, CNTN5 \n",
      "Negative:  TRG-AS1, ZNF665, PRPH, AC079465.1, AC021231.1, SLFN12L, DOCK8, AP002495.2, AC018742.1, S100A10 \n",
      "\t   LINC01933, PRDM6, SLC5A7, AC078785.1, SCMH1, SPP1, CD163, COL8A1, AC007388.1, GSE1 \n",
      "\t   LINC01122, MFSD1, CHST11, PAPOLG, UTS2, CHODL, AGAP1, BACH1, IGFBP7, AC109466.1 \n",
      "PC_ 3 \n",
      "Positive:  LINC01933, AC018742.1, AP002495.2, LINC00310, HMGN5, CEMIP2, SLC6A5, RASGEF1B, AC005154.6, LINC01122 \n",
      "\t   SH3RF1, AC091078.1, AC079465.1, AC009975.1, DLC1, HOMER1, CDH19, DTWD2, AC026316.5, NUP214 \n",
      "\t   NIPSNAP1, ADAMTS9, ASXL1, TNRC6B, NAV3, TIGAR, TMEM117, E2F5, CATSPER2, CFAP47 \n",
      "Negative:  AC008074.1, RUNX1, PRRX1, AL445250.1, XIST, EIF3L, AL049840.1, RPS6KB1, SPANXA2-OT1, AC008758.4 \n",
      "\t   DCX, BICDL1, PROX1, TASP1, RBFOX1, AC080162.1, ACOT9, ELMOD1, LINC00476, ELL \n",
      "\t   AL158071.4, SNX29, SLFN12L, PLEKHA6, GALNTL6, OSMR-AS1, IGF1R, MAU2, CSMD2, ZNF345 \n",
      "PC_ 4 \n",
      "Positive:  LINC01933, LINC00310, COL19A1, TRG-AS1, ROBO1, PALM2, EXT1, AC079465.1, GRIP1, AC244021.1 \n",
      "\t   DCX, AC019131.1, ZNF487, PCDH15, LINC01250, AC008537.1, AC091078.1, CPLANE1, NAV3, DOCK2 \n",
      "\t   AC018742.1, TMEM178B, PDGFD, ASXL1, TIGAR, STX7, SORBS2, SZRD1, C16orf87, MFSD1 \n",
      "Negative:  AL049840.1, STK35, AL031280.1, AC010636.2, LINC02240, FRAS1, HSPA12A, INPP4A, TANC2, SORCS2 \n",
      "\t   UCHL1, NEFH, CENPF, ANKMY2, LINC02082, AC008438.1, ARMH3, GHITM, TUBB2A, MANBA \n",
      "\t   TTK, CPNE8, E2F5, ACOT9, NIPAL2, KIF5A, COMMD6, COL8A1, AC124854.1, NRG1 \n",
      "PC_ 5 \n",
      "Positive:  SYT1, SCG2, ATP2B4, ASXL1, SLC17A6, TMEM163, PENK, CNR1, TAC1, SCN2A \n",
      "\t   RIOK1, IGFBP5, RELN, AC008537.1, GUCY1B1, SYT4, GAD1, ARF1, CDH8, ARHGEF28 \n",
      "\t   GRIA1, SLFN12L, CAAP1, CBLB, CALY, LINC01933, LINC01482, ZNF512, CARTPT, AC104662.2 \n",
      "Negative:  SLC6A5, AL356124.1, INA, CHRNA2, SLC32A1, RUNX1T1, NEFH, SPARCL1, CACNA1C, FAM117B \n",
      "\t   CNNM1, LAMA2, MORN4, UBR5, ANO4, PIGB, TRPC3, GULP1, AC107068.2, HOMER1 \n",
      "\t   HEPH, AC009495.1, AL139294.1, STEAP2, SLMAP, MAP9, MSI2, ZNF365, RUNX1, ERC2 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "debris.droplets.integrated.neurites <- FindVariableFeatures(debris.droplets.integrated.neurites)\n",
    "debris.droplets.integrated.neurites <- RunPCA(debris.droplets.integrated.neurites)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 2005\n",
      "Number of edges: 105217\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.5901\n",
      "Number of communities: 10\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".”\n"
     ]
    }
   ],
   "source": [
    "debris.droplets.integrated.neurites <- FindNeighbors(debris.droplets.integrated.neurites, dims = 1:20)\n",
    "debris.droplets.integrated.neurites <- FindClusters(debris.droplets.integrated.neurites, resolution = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "16:39:06 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "16:39:06 Read 2005 rows and found 20 numeric columns\n",
      "\n",
      "16:39:06 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "16:39:06 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:39:06 Writing NN index file to temp file /tmp/RtmpMBkzmG/file646350d59651\n",
      "\n",
      "16:39:06 Searching Annoy index using 8 threads, search_k = 3000\n",
      "\n",
      "16:39:07 Annoy recall = 100%\n",
      "\n",
      "16:39:08 Commencing smooth kNN distance calibration using 8 threads\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:39:12 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "16:39:12 Commencing optimization for 500 epochs, with 82886 positive edges\n",
      "\n",
      "16:39:19 Optimization finished\n",
      "\n"
     ]
    }
   ],
   "source": [
    "debris.droplets.integrated.neurites <- RunUMAP(debris.droplets.integrated.neurites, dims = 1:20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "neurite_debris_umap <- DimPlot(debris.droplets.integrated.neurites, label=T, raster = F, pt.size = 1.5)  + NoLegend()\n",
    "neurite_debris_umap <- rasterize(neurite_debris_umap, layers='Point', dpi=500)\n",
    "ggsave(filename = paste(fig_dir,'neurite_debris_umap.svg', sep='/'),\n",
    "       plot= neurite_debris_umap, device = 'svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "DefaultAssay(debris.droplets.integrated.neurites) <- 'RNA'\n",
    "\n",
    "neurite_debris_featureplot <- FeaturePlot(debris.droplets.integrated.neurites, pt.size = 1,\n",
    "                                          features=c('SLC17A6', 'SLC6A5', 'SLC5A7', 'GAD1'), label=T) \n",
    "neurite_debris_featureplot <- rasterize(neurite_debris_featureplot, layers='Point', dpi=500)\n",
    "\n",
    "ggsave(filename = paste(fig_dir,'neurite_debris_featureplot.svg', sep='/'),\n",
    "       plot= neurite_debris_featureplot, device = 'svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "neurite.exp.mat <- AverageExpression(debris.droplets.integrated.neurites, assays = 'RNA')$RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "marker.heatmap <- neurite.exp.mat[c('SLC5A7', 'SLC17A6', \n",
    "                                    'SLC6A5', 'GAD1', 'ACLY','GAD2'),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "marker.heatmap <- (marker.heatmap - rowMins(marker.heatmap))/(rowMaxs(marker.heatmap) - rowMins(marker.heatmap))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(marker.heatmap) <- paste('Cluster', colnames(marker.heatmap), sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# anno <- data.frame('cluster'= paste('cluster', colnames(marker.heatmap)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# colpal <- hue_pal()(10)\n",
    "# names(colpal) <- anno[,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "neurite_subclustering_heatmap <- \n",
    "    pheatmap(marker.heatmap, cluster_rows = TRUE, cluster_cols = TRUE, \n",
    "             filename = paste(fig_dir, 'neurite_subclustering_heatmap.pdf', sep='/'), \n",
    "             width = 6.4, height = 5.5, fontsize = 14) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# svglite(filename = paste(fig_dir, 'neurite_subclustering.svg', sep='/'))\n",
    "# pheatmap(marker.heatmap,  scale='row', annotation_col = anno, show_colnames = FALSE, annotation_colors = list(cluster=colpal))\n",
    "\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "cholinergic_glutamatergic_scatter <- FeatureScatter(debris.droplets.integrated.neurites, pt.size=0.2, feature1=c('SLC5A7'), feature2='SLC17A6')\n",
    "cholinergic_glutamatergic_scatter <- rasterize(cholinergic_glutamatergic_scatter, layer= 'Point', dpi=500) + \n",
    "    NoLegend() + FontSize(x.title=7, y.title=7, x.text=7, y.text=7) + theme(plot.title=element_blank())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cholinergic_glycinergic_scatter <- FeatureScatter(debris.droplets.integrated.neurites, pt.size = 0.2, feature1=c('SLC5A7'), feature2='SLC6A5')\n",
    "cholinergic_glycinergic_scatter <- rasterize(cholinergic_glycinergic_scatter, layer= 'Point', dpi=500) + \n",
    "    NoLegend() + FontSize(x.title=7, y.title=7, x.text=7, y.text=7) + theme(plot.title=element_blank())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cholinergic_gabaergic_scatter <- FeatureScatter(debris.droplets.integrated.neurites, pt.size=0.2, feature1=c('SLC5A7'), feature2='GAD1')\n",
    "cholinergic_gabaergic_scatter <- rasterize(cholinergic_gabaergic_scatter, layer= 'Point', dpi=500) +\n",
    "    NoLegend() + FontSize(x.title=7, y.title=7, x.text=7, y.text=7) + theme(plot.title=element_blank())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ggsave(filename = paste(fig_dir,'cholinergic_glutamatergic_scatter.svg', sep='/'),\n",
    "       plot= cholinergic_glutamatergic_scatter, device = 'svg', width=1, height=1, unit= 'in')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ggsave(filename = paste(fig_dir,'cholinergic_glycinergic_scatter.svg', sep='/'),\n",
    "       plot= cholinergic_glycinergic_scatter, device = 'svg', width=1, height=1, unit= 'in')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ggsave(filename = paste(fig_dir,'cholinergic_gabaergic_scatter.svg', sep='/'),\n",
    "       plot= cholinergic_gabaergic_scatter, device = 'svg', width=1, height=1, unit= 'in')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "saveRDS(debris.droplets.integrated.neurites, file= '/oak/stanford/groups/agitler/Shared/Matters_Arising/Gautier_Yadav/debris_droplets_integrated_neurites.RDS')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "debris.droplets.integrated.neurites <- readRDS(file= '/oak/stanford/groups/agitler/Shared/Matters_Arising/Gautier_Yadav/debris_droplets_integrated_neurites.RDS')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.1.1",
   "language": "R",
   "name": "ir411"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
